In [31]:
! pip install -r requirements.txt



   ---------------------------------------- 0.0/8.1 MB ? eta -:--:--
   ------------ --------------------------- 2.6/8.1 MB 13.4 MB/s eta 0:00:01
   -------------------------------- ------- 6.6/8.1 MB 16.2 MB/s eta 0:00:01
   ---------------------------------------- 8.1/8.1 MB 15.6 MB/s  0:00:00
   ---------------------------------------- 0.0/8.3 MB ? eta -:--:--
   ------------------------ --------------- 5.0/8.3 MB 24.4 MB/s eta 0:00:01
   ---------------------------------------- 8.3/8.3 MB 19.6 MB/s  0:00:00
   ---------------------------------------- 0.0/39.4 MB ? eta -:--:--
   ---- ----------------------------------- 4.7/39.4 MB 23.8 MB/s eta 0:00:02
   ---------- ----------------------------- 10.0/39.4 MB 23.7 MB/s eta 0:00:02
   ----------------- ---------------------- 16.8/39.4 MB 26.1 MB/s eta 0:00:01
   ----------------------- ---------------- 23.1/39.4 MB 26.9 MB/s eta 0:00:01
   ------------------------------ --------- 30.1/39.4 MB 28.1 MB/s eta 0:00:01
   ----------------


[notice] A new release of pip is available: 25.2 -> 25.3
[notice] To update, run: python.exe -m pip install --upgrade pip


### Importation des bibliothèques

In [33]:
import numpy as np 
import pandas as pd
from ucimlrepo import fetch_ucirepo

### Importation de la base

In [34]:
  
# fetch dataset 
cdc_diabetes_health_indicators = fetch_ucirepo(id=891) 
  
# data (as pandas dataframes) 
X = cdc_diabetes_health_indicators.data.features  ## Il s'agit de tous les features de la base
y = cdc_diabetes_health_indicators.data.targets   ## Il s'agit de la variable cible (avoir le diabète ou pas)
  

### Visualisation des metadonnées

In [35]:
pd.set_option("display.max_colwidth", None)
vars_meta = cdc_diabetes_health_indicators.variables
vars_meta[["name", "type", "description"]]


,name,type,description
0,ID,Integer,Patient ID
1,Diabetes_binary,Binary,0 = no diabetes 1 = prediabetes or diabetes
2,HighBP,Binary,0 = no high BP 1 = high BP
3,HighChol,Binary,0 = no high cholesterol 1 = high cholesterol
4,CholCheck,Binary,0 = no cholesterol check in 5 years 1 = yes cholesterol check in 5 years
5,BMI,Integer,Body Mass Index
6,Smoker,Binary,Have you smoked at least 100 cigarettes in your entire life? [Note: 5 packs = 100 cigarettes] 0 = no 1 = yes
7,Stroke,Binary,(Ever told) you had a stroke. 0 = no 1 = yes
8,HeartDiseaseorAttack,Binary,coronary heart disease (CHD) or myocardial infarction (MI) 0 = no 1 = yes
9,PhysActivity,Binary,physical activity in past 30 days - not including job 0 = no 1 = yes


In [3]:
X.head()

,HighBP,HighChol,CholCheck,BMI,Smoker,Stroke,HeartDiseaseorAttack,PhysActivity,Fruits,Veggies,...,AnyHealthcare,NoDocbcCost,GenHlth,MentHlth,PhysHlth,DiffWalk,Sex,Age,Education,Income
0,1,1,1,40,1,0,0,0,0,1,...,1,0,5,18,15,1,0,9,4,3
1,0,0,0,25,1,0,0,1,0,0,...,0,1,3,0,0,0,0,7,6,1
2,1,1,1,28,0,0,0,0,1,0,...,1,1,5,30,30,1,0,9,4,8
3,1,0,1,27,0,0,0,1,1,1,...,1,0,2,0,0,0,0,11,3,6
4,1,1,1,24,0,0,0,1,1,1,...,1,0,2,3,0,0,0,11,5,4


In [4]:
y.describe()

,Diabetes_binary
count,253680.000000
mean,0.139333
std,0.346294
min,0.000000
25%,0.000000
50%,0.000000
75%,0.000000
max,1.000000


In [ ]:
# Ajout de la cible y à X 
X['Diabetes_binary'] = y['Diabetes_binary'].values
X.head()

C:\Users\user\AppData\Local\Temp\ipykernel_22740\679768990.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X['Diabetes_binary'] = y['Diabetes_binary'].values


,HighBP,HighChol,CholCheck,BMI,Smoker,Stroke,HeartDiseaseorAttack,PhysActivity,Fruits,Veggies,...,NoDocbcCost,GenHlth,MentHlth,PhysHlth,DiffWalk,Sex,Age,Education,Income,Diabetes_binary
0,1,1,1,40,1,0,0,0,0,1,...,0,5,18,15,1,0,9,4,3,0
1,0,0,0,25,1,0,0,1,0,0,...,1,3,0,0,0,0,7,6,1,0
2,1,1,1,28,0,0,0,0,1,0,...,1,5,30,30,1,0,9,4,8,0
3,1,0,1,27,0,0,0,1,1,1,...,0,2,0,0,0,0,11,3,6,0
4,1,1,1,24,0,0,0,1,1,1,...,0,2,3,0,0,0,11,5,4,0


In [6]:
X.describe()

,HighBP,HighChol,CholCheck,BMI,Smoker,Stroke,HeartDiseaseorAttack,PhysActivity,Fruits,Veggies,...,NoDocbcCost,GenHlth,MentHlth,PhysHlth,DiffWalk,Sex,Age,Education,Income,Diabetes_binary
count,253680.000000,253680.000000,253680.000000,253680.000000,253680.000000,253680.000000,253680.000000,253680.000000,253680.000000,253680.000000,...,253680.000000,253680.000000,253680.000000,253680.000000,253680.000000,253680.000000,253680.000000,253680.000000,253680.000000,253680.000000
mean,0.429001,0.424121,0.962670,28.382364,0.443169,0.040571,0.094186,0.756544,0.634256,0.811420,...,0.084177,2.511392,3.184772,4.242081,0.168224,0.440342,8.032119,5.050434,6.053875,0.139333
std,0.494934,0.494210,0.189571,6.608694,0.496761,0.197294,0.292087,0.429169,0.481639,0.391175,...,0.277654,1.068477,7.412847,8.717951,0.374066,0.496429,3.054220,0.985774,2.071148,0.346294
min,0.000000,0.000000,0.000000,12.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000,1.000000,1.000000,1.000000,0.000000
25%,0.000000,0.000000,1.000000,24.000000,0.000000,0.000000,0.000000,1.000000,0.000000,1.000000,...,0.000000,2.000000,0.000000,0.000000,0.000000,0.000000,6.000000,4.000000,5.000000,0.000000
50%,0.000000,0.000000,1.000000,27.000000,0.000000,0.000000,0.000000,1.000000,1.000000,1.000000,...,0.000000,2.000000,0.000000,0.000000,0.000000,0.000000,8.000000,5.000000,7.000000,0.000000
75%,1.000000,1.000000,1.000000,31.000000,1.000000,0.000000,0.000000,1.000000,1.000000,1.000000,...,0.000000,3.000000,2.000000,3.000000,0.000000,1.000000,10.000000,6.000000,8.000000,0.000000
max,1.000000,1.000000,1.000000,98.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,...,1.000000,5.000000,30.000000,30.000000,1.000000,1.000000,13.000000,6.000000,8.000000,1.000000


##  Exploration de la base de données

In [8]:
X.shape


(253680, 22)

La base contient 253680 observations dont 22 colonnes (y compris la cible).

In [10]:
X["Diabetes_binary"].value_counts()


Diabetes_binary
0    218334
1     35346
Name: count, dtype: int64

In [12]:
X["Diabetes_binary"].value_counts(normalize=True) * 100


Diabetes_binary
0    86.066698
1    13.933302
Name: proportion, dtype: float64

La répartition des modalités de la variable cible est très déséquilibrée. Seulement 14% des individus sont diabétiques pour 86% qui ne le sont pas. Nous verrons plus loin comment nous exploitons les données pour faire face à cela.

In [13]:
X.isna().sum()

HighBP                  0
HighChol                0
CholCheck               0
BMI                     0
Smoker                  0
Stroke                  0
HeartDiseaseorAttack    0
PhysActivity            0
Fruits                  0
Veggies                 0
HvyAlcoholConsump       0
AnyHealthcare           0
NoDocbcCost             0
GenHlth                 0
MentHlth                0
PhysHlth                0
DiffWalk                0
Sex                     0
Age                     0
Education               0
Income                  0
Diabetes_binary         0
dtype: int64

Aucune valeur manquante détectée

In [16]:
np.isinf(X).sum()

HighBP                  0
HighChol                0
CholCheck               0
BMI                     0
Smoker                  0
Stroke                  0
HeartDiseaseorAttack    0
PhysActivity            0
Fruits                  0
Veggies                 0
HvyAlcoholConsump       0
AnyHealthcare           0
NoDocbcCost             0
GenHlth                 0
MentHlth                0
PhysHlth                0
DiffWalk                0
Sex                     0
Age                     0
Education               0
Income                  0
Diabetes_binary         0
dtype: int64

In [17]:
X.duplicated().sum()

np.int64(24206)

On a là 24206 doublons. A priori il s'agt d'observations dupliquées correspondant à des individus distincts. Elles seront donc conservées afin de préserver la distribution empirique des données et d’éviter l’introduction d’un biais artificiel.

In [ ]:
## Quelques doublons

X[X.duplicated(keep=False)].sort_values(by=X.columns.tolist())


,HighBP,HighChol,CholCheck,BMI,Smoker,Stroke,HeartDiseaseorAttack,PhysActivity,Fruits,Veggies,...,NoDocbcCost,GenHlth,MentHlth,PhysHlth,DiffWalk,Sex,Age,Education,Income,Diabetes_binary
4517,0,0,0,17,0,0,0,1,1,1,...,0,1,0,0,0,0,8,6,8,0
207307,0,0,0,17,0,0,0,1,1,1,...,0,1,0,0,0,0,8,6,8,0
42369,0,0,0,18,0,0,0,1,1,1,...,0,1,0,0,0,0,10,6,8,0
108949,0,0,0,18,0,0,0,1,1,1,...,0,1,0,0,0,0,10,6,8,0
17475,0,0,0,19,0,0,0,1,1,1,...,0,1,0,0,0,0,4,6,8,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
246364,1,1,1,40,0,0,0,1,1,1,...,0,2,0,0,0,0,9,6,7,0
99426,1,1,1,40,0,0,0,1,1,1,...,0,3,0,0,0,0,9,6,7,0
157792,1,1,1,40,0,0,0,1,1,1,...,0,3,0,0,0,0,9,6,7,0
41565,1,1,1,43,0,0,0,1,0,1,...,0,3,0,0,0,0,8,6,8,0


### Modalités de chaque variable 

In [21]:
for col in X.columns:
    print(col, ":", sorted(X[col].dropna().unique()))


HighBP : [np.int64(0), np.int64(1)]
HighChol : [np.int64(0), np.int64(1)]
CholCheck : [np.int64(0), np.int64(1)]
BMI : [np.int64(12), np.int64(13), np.int64(14), np.int64(15), np.int64(16), np.int64(17), np.int64(18), np.int64(19), np.int64(20), np.int64(21), np.int64(22), np.int64(23), np.int64(24), np.int64(25), np.int64(26), np.int64(27), np.int64(28), np.int64(29), np.int64(30), np.int64(31), np.int64(32), np.int64(33), np.int64(34), np.int64(35), np.int64(36), np.int64(37), np.int64(38), np.int64(39), np.int64(40), np.int64(41), np.int64(42), np.int64(43), np.int64(44), np.int64(45), np.int64(46), np.int64(47), np.int64(48), np.int64(49), np.int64(50), np.int64(51), np.int64(52), np.int64(53), np.int64(54), np.int64(55), np.int64(56), np.int64(57), np.int64(58), np.int64(59), np.int64(60), np.int64(61), np.int64(62), np.int64(63), np.int64(64), np.int64(65), np.int64(66), np.int64(67), np.int64(68), np.int64(69), np.int64(70), np.int64(71), np.int64(72), np.int64(73), np.int64(74)

Les variables sont toutes numériques. Certaines variables multimodales comme education, age, income sont ordonnées. Donc nous avons jugé qu'il n'y a pas de raison de les diviser avec le One-hot incoder. 